### PR TITLE
New lint `if_then_panic`

### LINKED ISSUE #7645: New lint: Manual assert
### What it does

Detects `if`-then-`panic!` that can be replaced with `assert!`

### Categories (optional)

- Kind: style

*What is the advantage of the recommended code over the original code*

It is simpler

### Drawbacks

None.

### Example

```rust
if !sad_people.is_empty() {
    panic!("there are sad people: {:?}", sad_people);
}
```

Could be written as:

```rust
assert!(sad_people.is_empty(), "there are sad people: {:?}", sad_people);
```

Lint name: `manual_assert` or `if_then_panic`

### PR DESCRIPTION
**Pull request by Labelray**
changelog: add the new lint [`if_then_panic`]
fix #7645

### CODE REVIEW

**Comment by camsteffen** [comment]
Use `cx.sess().source_map().is_multiline(span)` to make sure the condition fits on on line. Add a negative test for this.

I think providing a code suggestion is important for this lint. The fix may not be intuitive and people may not be aware that `assert!` accepts a message and formatting args. We probably need `PanicMacroExpn` similar to `FormatExpn`.

File: clippy_utils/src/higher.rs [hunk 1]
```diff
@@ -608,3 +608,106 @@ pub fn is_from_for_desugar(local: &hir::Local<'_>) -> bool {
 
     false
 }
+
+/// A parsed `panic!` expansion
+pub struct PanicExpn<'tcx> {
+    /// Span of `panic!(..)`
+    pub call_site: Span,
+    /// Inner `format_args!` expansion
+    pub format_args: PanicArgsExpn<'tcx>,
```
Comment (camsteffen) [inline clippy_utils/src/higher.rs:617]: We should reuse `FormatArgsExpn` here.

**Comment by camsteffen** [comment]
I'll leave it to @giraffate do to a full review. But note I still don't see the multi-line check. @Labelray don't hesitate to ask if something doesn't make sense!

**Comment by Labelray** [comment]
> I'll leave it to @giraffate do to a full review. But note I still don't see the multi-line check. @Labelray don't hesitate to ask if something doesn't make sense!

@camsteffen I didn't check the number of lines but I checked the number of statements. I think it is better.

**Comment by camsteffen** [comment]
A long condition can be a single statement.

```rust
if some_function(foo, bar, baz) == another_function(hi, whats, up)
    && some_function(a, b, c) == another_function(a, b, c) {
    panic!();
}
```

File: clippy_lints/src/if_then_panic.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                };
+                let mut applicability = Applicability::MachineApplicable;
+                let sugg = snippet_with_applicability(cx, span, "..", &mut applicability);
+
+                let cond_sugg =
+                if let ExprKind::DropTemps(Expr{kind: ExprKind::Unary(UnOp::Not, not_expr), ..}) = cond.kind {
+                    snippet_with_applicability(cx, not_expr.span, "..", &mut applicability).to_string()
+                } else {
+                    format!("!{}", snippet_with_applicability(cx, cond.span, "..", &mut applicability))
+                };
```
Comment (giraffate) [inline clippy_lints/src/if_then_panic.rs:83]: nits: Doesn't rustfmt work?
```suggestion
                let cond_sugg = if let ExprKind::DropTemps(Expr{kind: ExprKind::Unary(UnOp::Not, not_expr), ..}) = cond.kind {
                    snippet_with_applicability(cx, not_expr.span, "..", &mut applicability).to_string()
                } else {
                    format!("!{}", snippet_with_applicability(cx, cond.span, "..", &mut applicability))
                };
```

File: clippy_lints/src/if_then_panic.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+                span_lint_and_help(
+                    cx,
+                    IF_THEN_PANIC,
+                    expr.span,
+                    "only a `panic!` in `if`-then statement",
+                    None,
+                    &format!("considering use `assert!({}, {})` instead", cond_sugg, sugg)
+                );
```
Comment (giraffate) [inline clippy_lints/src/if_then_panic.rs:92]: Can this lint have a suggestion using `span_lint_and_sugg` method? The `// run-rustfix` line needs to be added to the beginning of the test file too.

File: clippy_lints/src/if_then_panic.rs (hunk 2 above)
Comment (Labelray) [reply clippy_lints/src/if_then_panic.rs:83]: I wrote the code as your suggest at first. But rustfmt generated a mistake.
```
error[internal]: line formatted, but exceeded maximum width (maximum: 120 (see `max_width` option), found: 126)
  --> /Users/labelray/Projects/rust-clippy/clippy_lints/src/if_then_panic.rs:78:78:121
   |
78 |                 let cond_sugg = if let ExprKind::DropTemps(Expr{kind: ExprKind::Unary(UnOp::Not, not_expr), ..}) = cond.kind {
   |                                                                                                                         ^^^^^^
   |

warning: rustfmt has failed to format. See previous 1 errors.
```

File: tests/ui/fallible_impl_from.stderr [hunk 4]
```diff
@@ -40,6 +40,16 @@ LL |             panic!();
...
+error: only a `panic!` in `if`-then statement
+  --> $DIR/fallible_impl_from.rs:28:9
+   |
+LL | /         if i != 42 {
+LL | |             panic!();
+LL | |         }
+   | |_________^ help: try: `assert!(!i != 42, "explicit panic");`
+   |
+   = note: `-D clippy::if-then-panic` implied by `-D warnings`
+
```
Comment (giraffate) [inline tests/ui/fallible_impl_from.stderr:52]: Can you add `#![allow(clippy::if_then_panic)]` and remove these warnings in this test file? It looks a little confusing.

**Review by giraffate — COMMENTED** [review]
I made a comment but overall looks good, thanks!

**Review by giraffate — APPROVED** [review]
Can the conflict be resolved?

**Comment by giraffate** [comment]
We follow [a no merge-commit policy](https://github.com/rust-lang/rust-clippy/blob/master/doc/basics.md#pr). Can you do a rebase, not a merge?

**Comment by giraffate** [comment]
@bors r+

Thanks!

File: tests/ui/fallible_impl_from.stderr (hunk 4 above)
Comment (pravic) [reply tests/ui/fallible_impl_from.stderr]: If this suggestion even valid?

`!i != 42` is `(!i) != 42` which is not what is intended to be suggested here. It should be `!(i != 42)`.

**Comment by camsteffen** [comment]
Hi @Jasperav, thanks for the feedback.

> This issue causes regression: `panic!` can be catch with `std::panic::catch_unwind`, but `assert!` not.

I don't think this is true, unless I misunderstood. See [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a5b53124d61639ccfb86776667f39570).

> I would suggest reverting this lint or don't make it active by default

The lint is being moved to pedantic (off by default) in #7810.

> which... is weird. Why negate an integer?

This should be fixed from #7741.

**Pull request merged**